Introduce short-lived `setup_taker::Actor`

luckysori opened the pull request:
Is this close to what you had in mind? I would welcome some async guidance in this time of scarce internet access.

There's at least a couple of things left to iron out, so I've left some TODOs etc. which I intend to delete before merging.

More importantly, the actor tests fail, so there's something crucial still missing in this implementation.

luckysori committed:
Introduce short-lived setup_taker_actor

thomaseizinger commented on daemon/src/setup_taker.rs:
```diff
@@ -0,0 +1,213 @@
+            .get_announcement
+            .send(oracle::GetAnnouncement(cfd.order.oracle_event_id))
+            .await?
+            .with_context(|| format!("Announcement {} not found", cfd.order.oracle_event_id))?;
```
Could we pass the announcement in on constructing the actor? That would remove `get_announcement` channel.

thomaseizinger commented on daemon/src/setup_taker.rs:
```diff
@@ -0,0 +1,213 @@
+            .send(oracle::MonitorAttestation {
+                event_id: offer_announcement.id,
+            })
+            .await?;
```
Can we do as within `taker_cfd::Actor` once it setup finishes?

thomaseizinger reviewed (commented):
Yep that looks really good! Can't spot anything obvious why the tests would be failing ...

luckysori commented on daemon/src/connection.rs:
```diff
@@ -25,6 +27,8 @@ pub struct Actor {
     /// Max duration since the last heartbeat until we die.
     timeout: Duration,
     connected_state: Option<ConnectedState>,
+    // QUESTION: Does this need to be generic?
```
:question: Is it okay to set `wallet::Actor` here?

thomaseizinger replied:
No I think that will become an issue once you try and insert something into this hashmap.

Best to remove the type parameter from `setup_contract` and pass two `MessageChannel`s in. If that works, try and land that as a separate patch :)

luckysori commented on daemon/src/setup_taker.rs:
```diff
@@ -0,0 +1,213 @@
+        //   `update_sender.send(db::load_all_cfds(conn).await?)?;`
+        // I see two options:
+        // - make `update_sender` a mpmc (via `crossbeam` for example).
+        // - send a message to the `taker_cfd` to update the feed.
```
:question: What do you think about this problem, @thomaseizinger?

thomaseizinger replied:
Can we move the update back into `taker_cfd::Actor` and you just send it a message back? Will require another `MessageChannel` but for now, that is probably the least invasive thing.

luckysori committed:
Prefer 2 MessageChannels over generic Address

luckysori committed:
Associate contract setup messages with an order ID

This is the first step in allowing concurrent execution of contract
setup for different CFDs.

luckysori committed:
Correct typo in actor tests

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -329,6 +266,85 @@ impl<O, M, W> Actor<O, M, W> {
+        self.register_setup_actor.send(msg).await?;
+
+        Ok(())
+    }
```
This should not be necessary? You know the address of the actor when you create it.

luckysori replied:
Yes, you're right. In fact, I was already inserting it into the hashmap, just not calling `insert`. Don't let me work late at night.

thomaseizinger commented on daemon/src/setup_taker.rs:
```diff
@@ -0,0 +1,252 @@
+                quantity: self.quantity,
+            })
+            .await
+            .expect("connection actor to be up");
```
I would have used a different message and actor here.

If you take an address instead of a message channel to the `connection::Actor`, you can make a new message `StartContractSetup` or something that contains:

- Your own address
- The order ID
- The quantity

The connection actor can then send the first message on your behalf.

luckysori replied:
I considered something like what you describe at first, but decided against it because I thought that taking an `Address` to the `connection::Actor` would force us to change the `taker_cfd::Actor` in ways that would break the actor tests. Admittedly, I didn't give it a go, so maybe I was wrong about that. I'll give it a go now.

thomaseizinger reviewed (commented):
> @thomaseizinger: I tackled the improvements suggested and figured out what the bug was: I forgot to actually put anything in these two `setup_actors` hashmaps, so we could never get any messages from the maker to the corresponding `setup_taker::Actor`.
> 
> Looking into this bug made me think about a couple of things:
> 
>  I'm not too fond of using the `wire` message types for anything other than sending over the wire. I couldn't simply add the address of the `setup_taker::Actor` to the message that was already going to the `connection::Actor` because the same message needed to be serializable and relevant to the maker.

Agreed, I would introduce separate messages between the actors.

>  Isn't it a suspicious sign that we keep the same `setup_actors` state in two different actors? I may have followed your original branch a little blindly in that regard. Perhaps it would be better if only the `taker_cfd::Actor` kept this state and it was responsible for both ensuring that there's only one contract setup per order ID at any given moment and that messages from the maker go to the correct `setup_taker::Actor`.

While that might look better at the moment, it doesn't actually take responsibility away from the `taker_cfd::Actor`. We want to introduce _boundaries_ between the actors which means the `taker_cfd::Actor` cannot act as a proxy for messages.

The fact that the `connection` actor has to keep a HashMap of addresses is an artifact of us _not_ having a transport that supports multiplexing. If we would be able to open an isolated "channel" over the connection (for example https://docs.rs/quinn/0.8.0/quinn/struct.Connection.html#method.open_bi), then the connection actor would not need to act as a proxy but would only need to hand the channel to the setup actor and could otherwise stop caring about it.

That is what I want to get to long-term and for that, removing the proxying from the `taker_cfd::Actor` is a step forward :)

thomaseizinger reviewed (approved):
Nice! Thank you! :)

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -766,22 +720,39 @@ where
+            Failed { order_id, error } => {
+                log_error!(self.handle_contract_setup_failed(order_id, error))
+            }
+        }
```
Actually, there is one more thing.

Can we _not_ split `Completed` into individual handlers here? With our new model, we will pass the entire result to the aggregate and no longer make this decision in the actor.

thomaseizinger replied:
Sorry for only noticing this now!

luckysori replied:
@thomaseizinger @da-kami: I think this is what you guys meant, but the difference is pretty small, so maybe you were referring to something else: https://github.com/itchysats/itchysats/pull/684/commits/d7f7d8f91292e11edea561832aafb3e170816d59

thomaseizinger replied:
Nah that is perfect, thank you!

da-kami reviewed (approved):
Great work!
As Thomas said it'd help us integrating the new model if we don't split up the contract setup completed in the handler.

luckysori committed:
fixup! Introduce short-lived setup_taker_actor
